reindexer/reindex_worker/src/test/scala/uk/ac/wellcome/platform/reindex_worker/services/ReindexWorkerServiceTest.scala

```diff
@@ -0,0 +1,216 @@
+      hybridRecord.copy(version = hybridRecord.version + 1,
+                        reindexVersion = reindexJob.desiredVersion)
+    )
+
+    val invalidSqsMessage = SQSMessage(
+      subject = None,
+      body = toJson(reindexJob).get,
+      topic = "topic",
+      messageType = "message",
+      timestamp = "now"
+    )
+
+    val service = reindexWorkerService()
+
+    val future = service.processMessage(message = invalidSqsMessage)
+
+    whenReady(future) { _ =>
+      val actualRecords: List[HybridRecord] =
+        Scanamo.scan[HybridRecord](dynamoDbClient)(tableName).map(_.right.get)
+
+      actualRecords shouldBe expectedRecords
+    }
+  }
+
+  it(
+    "returns a failed Future if it cannot parse the SQS message as a ReindexJob") {
+    val sqsMessage = SQSMessage(
+      subject = None,
+      body = "<xml>What is JSON.</xl?>",
+      topic = "topic",
+      messageType = "message",
+      timestamp = "now"
+    )
+
+    val service = reindexWorkerService()
+
+    val future = service.processMessage(message = sqsMessage)
+
+    whenReady(future.failed) { result =>
+      result shouldBe a[GracefulFailureException]
+      result.getMessage should include(
+        "expected json value got < (line 1, column 1)")
+    }
+  }
+
+  it("returns a failed Future if the reindex job fails") {
+    val reindexJob = ReindexJob(
+      shardId = "sierra/333",
+      desiredVersion = 3
+    )
+
+    val sqsMessage = SQSMessage(
+      subject = None,
+      body = toJson(reindexJob).get,
+      topic = "topic",
+      messageType = "message",
+      timestamp = "now"
+    )
+
+    val service = reindexWorkerService(dynamoTableName = "does-not-exist")
+
+    val future = service.processMessage(message = sqsMessage)
+
+    whenReady(future.failed) { result =>
+      result shouldBe a[GracefulFailureException]
+      result.getMessage should include(
+        "Cannot do operations on a non-existent table")
+    }
+  }
+
+  it("returns a failed Future if the reindex job fails after a delay") {
+    val targetService = mock[ReindexService]
+    when(
+      targetService.runReindex(any[ReindexJob])(
+        any[SourcedDynamoFormatWrapper[HybridRecord]]))
+      .thenReturn(Future {
+        Thread.sleep(500)
+        throw new RuntimeException("This took too long")
+      })
+
+    val service = new ReindexWorkerService(
+      targetService = targetService,
+      reader = mock[SQSReader],
+      snsWriter = mock[SNSWriter],
+      system = actorSystem,
+      metrics = metricsSender
+    )
+
+    val reindexJob = ReindexJob(
+      shardId = "sierra/444",
+      desiredVersion = 4
+    )
+
+    val sqsMessage = SQSMessage(
+      subject = None,
+      body = toJson(reindexJob).get,
+      topic = "topic",
+      messageType = "message",
+      timestamp = "now"
+    )
+
+    val future = service.processMessage(message = sqsMessage)
+
+    whenReady(future.failed) { result =>
+      result shouldBe a[GracefulFailureException]
```

Comment:
I think it would be nicer to say `result shouldBe GracefulFailureException(exception)` where `exception` is `new RuntimeException("This took too long")`

Reply:
I have fixed this, but GitHub is resolutely not updating its diff 😕